pkg/kvstore/etcd.go

```diff
@@ -1111,6 +1122,39 @@ reList:
 	}
 }
 
+func (e *etcdClient) paginatedList(ctx context.Context, log *logrus.Entry, prefix string,
+	process func(*mvccpb.KeyValue)) (revision int64, err error) {
+	start, end := prefix, client.GetPrefixRangeEnd(prefix)
+
+	for {
+		res, err := e.client.Get(ctx, start, client.WithRange(end),
+			client.WithSort(client.SortByKey, client.SortAscend),
+			client.WithRev(revision), client.WithSerializable(),
+			client.WithLimit(int64(e.listBatchSize)),
+		)
+		if err != nil {
+			return 0, err
+		}
+
+		log.WithFields(logrus.Fields{
+			fieldNumEntries:       len(res.Kvs),
+			fieldRemainingEntries: res.Count - int64(len(res.Kvs)),
```

Comment:
AFAIU it is always returned by etcd \[1\], \[2\], and from an empirical test it works (elements = 10, limit = 4):

```
level=debug msg="Received list response from etcd" numEntries=4 remainingEntries=6 subsys=kvstore
level=debug msg="Received list response from etcd" numEntries=4 remainingEntries=2 subsys=kvstore
level=debug msg="Received list response from etcd" numEntries=2 remainingEntries=0 subsys=kvstore
```

\[1\]: https://github.com/etcd-io/etcd/blob/08407ff7600eb16c4445d5f21c4fafaf19412e24/server/etcdserver/apply.go#L418
\[2\]: https://github.com/etcd-io/etcd/blob/08407ff7600eb16c4445d5f21c4fafaf19412e24/server/mvcc/kvstore_txn.go#L179

Reply:
While checking the etcd source code, I also discovered that sorting is not performed by default if the `WithSort` option is not specified \[1\]. That shouldn't introduce a big penalty though, since pdqsort is optimized to complete in linear time when the input is already sorted \[2\], which is the case \[3\].

\[1\]: https://github.com/etcd-io/etcd/blob/08407ff7600eb16c4445d5f21c4fafaf19412e24/server/etcdserver/apply.go#L390-L410
\[2\]: https://github.com/orlp/pdqsort#the-best-case
\[3\]: https://github.com/etcd-io/etcd/blob/08407ff7600eb16c4445d5f21c4fafaf19412e24/server/etcdserver/apply.go#L385-L387